### PR TITLE
Honor association views when wrapping

### PR DESCRIPTION
**Pull request by ianks**
Previously, when wrapping a relation, the view for the association would not be honored. This would make the output of `.wrap` different from combine, and cause unexpected behavior.

This PR addresses that by first applying the view to the association target if there is one. 

Also, since `.wrap` with prefix attributes, any previously aliased attributes would fail loudly since `Sequel::SQL::AliasedExpression` does not respond to `.as`. To fix this, we now will handle alias `Sequel::SQL::AliasedExpression` by constructing a new instance with the new alias name.

### CODE REVIEW

**Comment by ianks** [comment]
Can i get a review on this?

File: lib/rom/sql/relation/reading.rb [hunk 1]
```diff
@@ -1025,6 +1025,31 @@ def nullify
...
+          others = names.map do |name|
+            association = associations[name]
+            opts = association.options
+            definition = association.definition
+            target = association.target
+            new_target = association.view ? target.send(association.view) : target
+            to_wrap = association.class.allocate
+            to_wrap.send(:initialize, definition, opts.merge(target: new_target))
+            to_wrap.wrap
+          end
```
Comment (solnic) [inline lib/rom/sql/relation/reading.rb]: This indicates that there's a missing abstraction. It seems like having a new method on association objects that will handle all of this (and provide a meaningful name at the same time) would be a good idea. Could you come up with something?

File: spec/integration/wrap_spec.rb [hunk 2]
```diff
@@ -86,6 +86,39 @@
...
         end
       end
+
+      context 'using association with a view' do
+        before do
+          conf.relation(:users) do
+            auto_map false
+
+            schema(infer: true) do
+            end
```
Comment (solnic) [inline spec/integration/wrap_spec.rb]: No need for this empty block

File: spec/integration/wrap_spec.rb [hunk 3]
```diff
@@ -86,6 +86,39 @@
...
+          end
+        end
+
+        it 'includes the extra attributes' do
+          result = tasks.wrap(:users).to_a
+
+          result.each do |task|
+            expect(task[:users]).to eql(extra_attribute: 'testing')
+          end
```
Comment (solnic) [inline spec/integration/wrap_spec.rb]: It would be better to materialize this relation and compare it with the expected resulting structure. If, for whatever reason, `result` is empty, that `each` won't do a thing and the spec will continue to pass.

File: spec/unit/attribute_spec.rb [hunk 4]
```diff
@@ -94,6 +94,13 @@
     end
   end
 
+  describe '#aliased' do
+    it 'can alias an previously aliased attribute' do
+      expect(users[:id].as(:uid).as(:uuid).sql_literal(ds)).
+        to eql(%("users"."id" AS "uuid"))
+    end
+  end
```
Comment (solnic) [inline spec/unit/attribute_spec.rb]: It would be good to extract this into a separate PR. I think it's a good fix that can be shipped separately.

File: lib/rom/sql/attribute.rb [hunk 5]
```diff
@@ -36,7 +36,7 @@ def self.[](*args)
       # @api public
       def aliased(alias_name)
         super.with(name: name || alias_name).meta(
-          sql_expr: sql_expr.as(alias_name)
+          sql_expr: alias_sql_expr(sql_expr, alias_name)
```
Comment (flash-gordon) [inline lib/rom/sql/attribute.rb]: it looks like we should have our own composable wrapper for aliased expressions, that case analysis in `alias_sql_expr` is leaking

**Comment by ianks** [comment]
@flash-gordon I noticed a weird error when wrapping an aliased function. essentially, the attribute name has an extra prefix that is never unwrapped, so for the failing spec https://github.com/rom-rb/rom-sql/pull/335/commits/27f46040ef5a21552dbc1ccd760991b543019b44, the wrapped key is `users_extra_attribute` rather than `extra_attribute`. Do you have any clues as to why this might happen? Stuck on it right now, but would like to get it fixed :)

**Comment by ianks** [comment]
Fixed the issue with aliasing functions. But now I've discovered a new regression in 5.0, aliasing attributes in wrapped relations does not work. So working on that now.

File: lib/rom/sql/attribute.rb (hunk 5 above)
Comment (ianks) [reply lib/rom/sql/attribute.rb]: I think this will be challenging to do in this PR, i'm not sure exactly how to approach it since from what i can see, `sql_expr` is an actual Sequel::* class everywhere so it may have unexpected breakage. Can we postpone this refactoring?

**Comment by ianks** [comment]
Alright, so I *believe* this PR is ready. It might not be the prettiest, but at least the behavior is correct now. With this patch, our (large) production app can now pass all specs on rom 5!

### What this patch addresses

- views are now honored in wrapped associations
- attribute aliasing on wrapped attributes used the correct alias name
- proper aliasing for functions with wrapped attributes

Excited to get this in so we do not have to run off a fork, and contribute upstream to bugfixes, features, etc.

- Ian

File: lib/rom/sql/relation/reading.rb [hunk 6]
```diff
@@ -1011,6 +1011,35 @@ def unfiltered
...
+        # @see http://sequel.jeremyevans.net/rdoc-plugins/files/lib/sequel/extensions/null_dataset_rb.html
+        # @example result will always be empty, regardless if records exists
+        #   users.where(name: 'Alice').nullify
+        #
+        # @return [SQL::Relation]
+        #
+        # @api public
+        def nullify
+          new(dataset.__send__(__method__))
+        end
```
Comment (solnic) [inline lib/rom/sql/relation/reading.rb]: This seems to be a left-over from another PR, right? 😄

File: lib/rom/sql/associations/core.rb [hunk 7]
```diff
@@ -14,6 +14,14 @@ def preload(target, loaded)
...
           target.where(target_key => target_pks)
         end
+
+        # @api private
+        def wrapped
+          new_target = view ? target.send(view) : target
+          to_wrap = self.class.allocate
+          to_wrap.send(:initialize, definition, options.merge(target: new_target))
+          to_wrap.wrap
+        end
```
Comment (solnic) [inline lib/rom/sql/associations/core.rb]: Could you remind me why you had to use `Object#allocate` here?

File: spec/unit/attribute_spec.rb [hunk 8]
```diff
@@ -94,6 +94,13 @@
     end
   end
 
+  describe '#aliased' do
+    it 'can alias an previously aliased attribute' do
```
Comment (psparrow) [inline spec/unit/attribute_spec.rb]: nit: "an" => "a"

File: lib/rom/sql/attribute_aliasing.rb [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+      # @example
+      #   users[:id].aliased(:user_id)
+      #
+      # @return [SQL::Attribute]
+      #
+      # @api public
+      def aliased(alias_name)
+        new_name, new_alias_name = extract_alias_names(alias_name)
+
+        # super.with(name: name || alias_name).meta(
```
Comment (psparrow) [inline lib/rom/sql/attribute_aliasing.rb]: cut commented code

File: lib/rom/sql/associations/core.rb (hunk 7 above)
Comment (ianks) [reply lib/rom/sql/associations/core.rb]: `Association#new` is overridden with defaults that do not work for this use case

**Comment by ianks** [comment]
PR comments have been addressed :)

File: lib/rom/sql/associations/core.rb (hunk 7 above)
Comment (solnic) [reply lib/rom/sql/associations/core.rb]: @ianks could we tweak the underlying implementation in a way that would not require using `allocate` + `send(:initialize, ...)` here? something like `new(target: view ? target.send(view) : target)` should work.

Comment (solnic) [reply lib/rom/sql/associations/core.rb]: ...sorry, I meant `with(target: view ? target.send(view).target)`, that's the conventional method we use in rom.

Comment (ianks) [reply lib/rom/sql/associations/core.rb]: does this require changes to rom-core? @solnic

Comment (solnic) [reply lib/rom/sql/associations/core.rb]: @ianks OK I see what's going on, it's due to the way `Associations::Core.new` is implemented. This is unfortunate, I'll do something about it later so for now let's leave this `allocate` trick here.

**Comment by solnic** [comment]
@flash-gordon could you take a look as well, please?

File: lib/rom/sql/mapper_compiler.rb [hunk 10]
```diff
@@ -7,11 +7,20 @@ def visit_attribute(node)
...
         else
           [name]
         end
       end
+
+      private
+
+      def extract_wrapped_name(node)
+        _, _, meta_options = node
+        unwrapped_name = meta_options[:alias].to_s
```
Comment (flash-gordon) [inline lib/rom/sql/mapper_compiler.rb]: in ruby 2.7, `Symbol#to_s` start returning frozen strings, this code will break

**Comment by ianks** [comment]
@flash-gordon @solnic fixed the last issue. code climate seems to be failing for an unrelated reason. merge? :smile:

**Comment by flash-gordon** [comment]
@ianks nice, please squash the commits, I'll merge it shortly after

**Comment by ianks** [comment]
@flash-gordon All squashed.
